### PR TITLE
Total coins endpoint shows stats for kyc passed users (+cache)

### PR DESCRIPTION
**Pull request by ice-cronus**
Filter users and its referral's balances using kyc passed = 2 step (face verified) for total coins stats
Due to more complex and slow query - add caching logic to keep values in redis (with 5 min expiration if not configured another option)

### CODE REVIEW

File: bookkeeper/storage/contract.go [hunk 1]
```diff
@@ -24,15 +24,17 @@ type (
 		Ping(ctx context.Context) error
 		Insert(ctx context.Context, columns *Columns, input InsertMetadata, usrs []*model.User) error
 		SelectBalanceHistory(ctx context.Context, id int64, createdAts []stdlibtime.Time) ([]*BalanceHistory, error)
-		SelectTotalCoins(ctx context.Context, createdAts []stdlibtime.Time) ([]*TotalCoins, error)
+		SelectTotalCoins(ctx context.Context, createdAts []stdlibtime.Time, kycStepToCalculateTotals uint8) ([]*TotalCoins, error)
```
Comment (ice-ares) [inline bookkeeper/storage/contract.go]: Not needed

File: bookkeeper/storage/contract.go [hunk 2]
```diff
@@ -24,15 +24,17 @@ type (
...
 	}
 	BalanceHistory struct {
 		CreatedAt                               *time.Time
 		BalanceTotalMinted, BalanceTotalSlashed float64
 	}
 	TotalCoins struct {
-		CreatedAt                                                          *time.Time
-		BalanceTotalStandard, BalanceTotalPreStaking, BalanceTotalEthereum float64
+		CreatedAt              *time.Time `redis:"created_at"`
```
Comment (ice-ares) [inline bookkeeper/storage/contract.go]: No. Cache the response, as json directly

File: tokenomics/balance.go [hunk 3]
```diff
@@ -36,9 +38,22 @@ func (r *repository) GetTotalCoinsSummary(ctx context.Context, days uint64, utcO
 		dates = append(dates, date)
 		res.TimeSeries = append(res.TimeSeries, &TotalCoinsTimeSeriesDataPoint{Date: date})
 	}
-	totalCoins, err := r.dwh.SelectTotalCoins(ctx, dates)
+	totalCoins, misses, err := r.cachedTotalCoins(ctx, dates)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to SelectTotalCoins for createdAts:%#v", dates)
+		misses = dates
+		log.Warn("%v", errors.Wrapf(err, "failed to fetch from totalCoinStats from cache, fetching from CH, misses: %v", misses))
```
Comment (ice-ares) [inline tokenomics/balance.go]: Please learn how to use that logging api

File: tokenomics/balance.go [hunk 4]
```diff
@@ -36,9 +38,22 @@ func (r *repository) GetTotalCoinsSummary(ctx context.Context, days uint64, utcO
...
+	}
+	if len(misses) > 0 {
+		extraCoins, err := r.dwh.SelectTotalCoins(ctx, misses, uint8(users.LivenessDetectionKYCStep))
```
Comment (ice-ares) [inline tokenomics/balance.go]: No. I dont mix dates, hit with missed dates, u cache everything as a whole. Based on the days argument

File: tokenomics/balance.go [hunk 5]
```diff
@@ -342,3 +357,63 @@ func ApplyPreStaking(amount, preStakingAllocation, preStakingBonus float64) (flo
...
+				}
+			}
+			if !found {
+				misses = append(misses, d)
+			}
+		}
+	}
+	return cached, misses, nil
+}
+func (r *repository) cacheTotalCoins(ctx context.Context, coins []*dwh.TotalCoins) error {
```
Comment (ice-ares) [inline tokenomics/balance.go]: No. Serialize response to json, set to a key

File: tokenomics/contract.go [hunk 6]
```diff
@@ -341,5 +341,6 @@ type (
 			Parent stdlibtime.Duration `yaml:"parent"`
 			Child  stdlibtime.Duration `yaml:"child"`
 		} `yaml:"globalAggregationInterval"`
+		CoinStatsCacheDuration stdlibtime.Duration `yaml:"coinStatsCacheDuration"`
```
Comment (ice-ares) [inline tokenomics/contract.go]: No. Use aggregation min/child

File: bookkeeper/storage/storage.go [hunk 7]
```diff
@@ -541,14 +541,32 @@ func (db *db) SelectTotalCoins(ctx context.Context, createdAts []stdlibtime.Time
 		format := date.UTC().Format(stdlibtime.RFC3339)
 		createdAtArray = append(createdAtArray, format[0:len(format)-1])
 	}
+	sql := fmt.Sprintf(`
+			SELECT u.created_at as created_at,
+				   sum((u.balance_solo + u.balance_t0 + verified_balance_t1.balance + verified_balance_t2.balance)*(100.0-u.pre_staking_allocation)/100.0)  AS balance_total_standard,
+				   sum((u.balance_solo + u.balance_t0 + verified_balance_t1.balance + verified_balance_t2.balance)* (100 + u.pre_staking_bonus) * u.pre_staking_allocation / 10000) AS balance_total_pre_staking,
```
Comment (ice-ares) [inline bookkeeper/storage/storage.go]: U didnt pay attention to the sql i made yesterday

**Review by ice-ares — CHANGES_REQUESTED** [review]

File: cmd/freezer-coin-distributer/api/docs.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: ice License 1.0
+
+// Package api Code generated by swaggo/swag. DO NOT EDIT
+package api
```
Comment (ice-ares) [inline cmd/freezer-coin-distributer/api/docs.go]: huh?

File: bookkeeper/storage/contract.go (hunk 2 above)
Comment (ice-ares) [reply bookkeeper/storage/contract.go]: roll this back

File: tokenomics/balance.go [hunk 9]
```diff
@@ -31,6 +32,11 @@ func (r *repository) GetTotalCoinsSummary(ctx context.Context, days uint64, utcO
 		truncationInterval                = r.cfg.GlobalAggregationInterval.Child
 		dayInterval                       = r.cfg.GlobalAggregationInterval.Parent
 	)
+	if cached, err := r.cachedTotalCoins(ctx, now, days); err == nil && cached != nil {
+		return cached, nil
+	} else if err != nil {
+		log.Error(errors.Wrapf(err, "failed to get coinStats from cache %v %v", now.Truncate(truncationInterval), days))
```
Comment (ice-ares) [inline tokenomics/balance.go]: return it

File: tokenomics/balance.go [hunk 10]
```diff
@@ -53,6 +59,9 @@ func (r *repository) GetTotalCoinsSummary(ctx context.Context, days uint64, utcO
 		child.Date = child.Date.In(location)
 	}
 	res.TotalCoins = res.TimeSeries[0].TotalCoins
+	if err = r.cacheTotalCoins(ctx, now, days, res); err != nil {
+		log.Error(errors.Wrapf(err, "failed to place cache for coinStats %v %v", now.Truncate(truncationInterval), days))
```
Comment (ice-ares) [inline tokenomics/balance.go]: return it

File: tokenomics/balance.go [hunk 11]
```diff
@@ -31,6 +32,11 @@ func (r *repository) GetTotalCoinsSummary(ctx context.Context, days uint64, utcO
 		truncationInterval                = r.cfg.GlobalAggregationInterval.Child
 		dayInterval                       = r.cfg.GlobalAggregationInterval.Parent
 	)
+	if cached, err := r.cachedTotalCoins(ctx, now, days); err == nil && cached != nil {
```
Comment (ice-ares) [inline tokenomics/balance.go]: getCachedTotalCoins

File: tokenomics/balance.go [hunk 12]
```diff
@@ -342,3 +351,33 @@ func ApplyPreStaking(amount, preStakingAllocation, preStakingBonus float64) (flo
 
 	return standardAmount, preStakingAmount
 }
+
+func (r *repository) cachedTotalCoins(ctx context.Context, now *time.Time, days uint64) (cached *TotalCoinsSummary, err error) {
+	cachedData, err := r.db.Get(ctx, coinStatsCacheKey(now.UTC().Truncate(r.cfg.GlobalAggregationInterval.Child), days)).Bytes()
```
Comment (ice-ares) [inline tokenomics/balance.go]: why are you trying to complicate things? Just days is enough. And add an expiry

Comment (ice-ares) [reply tokenomics/balance.go]: expiry = r.cfg.GlobalAggregationInterval.Child

File: tokenomics/balance.go [hunk 13]
```diff
@@ -342,3 +351,33 @@ func ApplyPreStaking(amount, preStakingAllocation, preStakingBonus float64) (flo
...
+	if err != nil && !errors.Is(err, redis.Nil) {
+		return nil, errors.Wrapf(err, "failed to get cached value for coinStats for %v %v", now, days)
+	}
+	if len(cachedData) == 0 {
+		return nil, nil
+	}
+	cached = new(TotalCoinsSummary)
+	if err = json.Unmarshal(cachedData, cached); err != nil {
```
Comment (ice-ares) [inline tokenomics/balance.go:364]: unmarshalContext, cuz u have *time.Time

File: tokenomics/balance.go [hunk 14]
```diff
@@ -342,3 +351,33 @@ func ApplyPreStaking(amount, preStakingAllocation, preStakingBonus float64) (flo
...
+		return errors.Wrapf(err, "failed to serialize cache value %v", summary)
+	}
+	expiration := r.cfg.GlobalAggregationInterval.Child
+
+	return errors.Wrapf(
+		r.db.SetNX(ctx, coinStatsCacheKey(now.UTC().Truncate(r.cfg.GlobalAggregationInterval.Child), days), cacheData, expiration).Err(),
+		"failed to save cache with coin stats (%v,%v) %+v", now, days, summary)
+}
+func coinStatsCacheKey(nowTruncated stdlibtime.Time, days uint64) string {
+	return fmt.Sprintf("coinStats:%v:%v", nowTruncated.Format(stdlibtime.RFC3339), days)
```
Comment (ice-ares) [inline tokenomics/balance.go]: just days

File: bookkeeper/storage/storage.go [hunk 15]
```diff
@@ -541,14 +541,39 @@ func (db *db) SelectTotalCoins(ctx context.Context, createdAts []stdlibtime.Time
 		format := date.UTC().Format(stdlibtime.RFC3339)
 		createdAtArray = append(createdAtArray, format[0:len(format)-1])
 	}
+	sql := fmt.Sprintf(`
+			SELECT u.created_at as created_at,
```
Comment (ice-ares) [inline bookkeeper/storage/storage.go]: embed it into a `select_total_coins.sql`

File: bookkeeper/storage/storage.go [hunk 16]
```diff
@@ -541,14 +541,39 @@ func (db *db) SelectTotalCoins(ctx context.Context, createdAts []stdlibtime.Time
...
+			GLOBAL LEFT JOIN
+				 (SELECT DISTINCT ON (id_tminus1, created_at) id_tminus1, created_at, sum(balance_for_tminus1) AS balance, sum(balance_for_tminus1_ethereum) AS ethereum
+				  FROM %[1]v
+				  WHERE created_at IN ['%[2]v']
+					AND kyc_step_passed >= %[3]v AND ( kyc_step_blocked = 0 OR kyc_step_blocked >= %[4]v)
+				 GROUP BY id_tminus1, created_at) AS verified_balance_t2 ON verified_balance_t2.id_tminus1 = u.id AND verified_balance_t2.created_at = u.created_at
+			WHERE u.created_at IN ['%[2]v']
+				AND u.kyc_step_passed >= %[3]v AND ( u.kyc_step_blocked = 0 OR u.kyc_step_blocked >= %[4]v)
+			GROUP BY u.created_at
+`, tableName, strings.Join(createdAtArray, "','"), kycStepToCalculateTotals, kycStepToCalculateTotals+1)
```
Comment (ice-ares) [inline bookkeeper/storage/storage.go]: do the +1 in sql

File: cmd/freezer-coin-distributer/api/docs.go (hunk 8 above)
Comment (ice-cronus) [reply cmd/freezer-coin-distributer/api/docs.go]: As I've seen checkgenerated failed on PR, but I can rollback it, u're not ok with that.

File: tokenomics/balance.go (hunk 13 above)
Comment (ice-cronus) [reply tokenomics/balance.go:364]: No, it is [stdlib.Time](https://github.com/ice-blockchain/freezer/blob/master/tokenomics/contract.go#L104), I tried unmarshalContext, it is failing due to missing unmarshaller

Comment (ice-ares) [reply tokenomics/balance.go:364]: oh, right

File: tokenomics/balance.go (hunk 12 above)
Comment (ice-cronus) [reply tokenomics/balance.go]: it is expirable https://github.com/ice-blockchain/freezer/blob/feature/total_coins_filtered/tokenomics/balance.go#L378

File: cmd/freezer-coin-distributer/api/docs.go (hunk 8 above)
Comment (ice-ares) [reply cmd/freezer-coin-distributer/api/docs.go]: always keep ur local master in sync with origin. Especially before u create a branch. Create it from origin/master, not master

Comment (ice-cronus) [reply cmd/freezer-coin-distributer/api/docs.go]: Yes, it works now, not sure what was the issue, but it was red althrough I've created branch from today's morning master

**Pull request merged**